### PR TITLE
Add Login pipeline

### PR DESCRIPTION
**Pull request by yuya-yuliya**
- add GetCommerceUser, IdentifyCommerceUser, MergeCarts, and SetCurrentUser processors
- add publishing profile in Folundation.Account project
- rename Wooli.Foundation.Account.config

### CODE REVIEW

File: src/Foundation/Commerce/code/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+namespace Wooli.Foundation.Commerce
+{
+    public static class Constants
+    {
+        public const string SignInRedirect = "/signIn";
```
Comment (stanislav-kozorez) [inline src/Foundation/Commerce/code/Constants.cs]: I have a feeling tha this shouldn't be a part of Foundation.Commerce

File: src/Foundation/Account/code/App_Config/Include/Foundation/Wooli/Foundation.Account.config [hunk 2]
```diff
@@ -4,7 +4,7 @@
   <sitecore>
     <pipelines>
       <wooli.foundation.account.login>
-        <processor type="Wooli.Foundation.Account.Infrastructure.Pipelines.Login, Wooli.Foundation.Account" resolve="true" />
+        <processor type="Wooli.Foundation.Account.Infrastructure.Pipelines.LoginProcessor, Wooli.Foundation.Account" resolve="true" />
```
Comment (evgenii-nikulin) [inline src/Foundation/Account/code/App_Config/Include/Foundation/Wooli/Foundation.Account.config]: Here part of namespace is missing
```suggestion
        <processor type="Wooli.Foundation.Account.Infrastructure.Pipelines.Login.LoginProcessor, Wooli.Foundation.Account" resolve="true" />
```

File: src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/GetCommerceUserProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                this.Abort(args);
+            }
+            else
+            {
+                this.userMapper.MapToLoginPipelineArgs(user, args);
+            }
+        }
+
+        private void Abort(LoginPipelineArgs args)
```
Comment (larionovadaria) [inline src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/GetCommerceUserProcessor.cs]: There's a property in PipelineArgs - Aborted. You can use it in controller

File: src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/GetCommerceUserProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        public GetCommerceUserProcessor(ICustomerProvider customerProvider, IUserMapper userMapper)
+        {
+            Assert.ArgumentNotNull(customerProvider, nameof(customerProvider));
+            Assert.ArgumentNotNull(userMapper, nameof(userMapper));
+
+            this.customerProvider = customerProvider;
+            this.userMapper = userMapper;
+        }
+
+        public override void Process(LoginPipelineArgs args)
```
Comment (larionovadaria) [inline src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/GetCommerceUserProcessor.cs]: - All the processors should be covered by unit tests. 

The following comments can be implemented in scope of the next request:
- Hide the logic of getting username under `this.customerProvider.GetCommerceUser`. For that create 2 methods in the customerProvider interface: `GetUser(string email)` and `GetUser(Guid contactId)`. Make `GetCommerceUser` obsolete, or remove if no references
- Do not forget about unit tests

File: src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/MergeCartsProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            this.cartService = cartService;
+        }
+
+        public override void Process(LoginPipelineArgs args)
+        {
+            Assert.ArgumentNotNull(args, nameof(args));
+
+            if (args.AnonymousContactId != null)
+            {
+                this.cartService.MergeCarts(args.AnonymousContactId);
```
Comment (larionovadaria) [inline src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/MergeCartsProcessor.cs]: Merge cart has return value, result can be `Success: false`

File: src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/SetCurrentUserProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+
+        public override void Process(LoginPipelineArgs args)
+        {
+            Assert.ArgumentNotNull(args, nameof(args));
+
+            args.AnonymousContactId = this.visitorContext.ContactId;
+
+            var user = this.userMapper.Map<LoginPipelineArgs, User>(args);
```
Comment (larionovadaria) [inline src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/SetCurrentUserProcessor.cs]: Why sometimes you use Map<> and sometimes MapTo...()?

File: src/Foundation/Commerce/code/ModelMappers/Users/UserMapper.cs [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    using Models.Mappings;
+
+    using Mapper = Base.Mappers.Mapper;
+
+    [Service(typeof(IUserMapper), Lifetime = Lifetime.Transient)]
+    public class UserMapper : Mapper, IUserMapper
+    {
+        public UserMapper()
+        {
+            var configuration = new MapperConfiguration(cfg => cfg.AddProfile<UserProfile>());
```
Comment (larionovadaria) [inline src/Foundation/Commerce/code/ModelMappers/Users/UserMapper.cs]: If we always do the same here, after agreed on using profiles, it makes sense to move logic to generic `Mapper<TProfile>`. Create a task for that in the next sprint.

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — CHANGES_REQUESTED** [review]

File: src/Feature/Account/code/Constants.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+// 
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+namespace Wooli.Feature.Account
+{
+    public static class Constants
```
Comment (larionovadaria) [inline src/Feature/Account/code/Constants.cs]: change to Constants.Redirects.SignIn

File: src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/IdentifyCommerceUserProcessor.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+            Assert.ArgumentNotNull(commerceTrackingService, nameof(commerceTrackingService));
+
+            this.commerceTrackingService = commerceTrackingService;
+        }
+
+        public override void Process(LoginPipelineArgs args)
+        {
+            Assert.ArgumentNotNull(args, nameof(args));
+
+            this.commerceTrackingService.IdentifyAs("CommerceUser", args.UserName);
```
Comment (larionovadaria) [inline src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/IdentifyCommerceUserProcessor.cs]: this method can return exception, e.g. someone decides to change login controller from mvc to webapi. Take care of exceprion handling

File: src/Foundation/Commerce/code/Models/Mappings/UserProfile.cs [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+            this.CreateMap<User, LoginPipelineArgs>()
+                .ForMember(dest => dest.UserName, opt => opt.MapFrom(src => src.UserName))
+                .ForMember(dest => dest.ContactId, opt => opt.MapFrom(src => src.ContactId))
+                .ForMember(dest => dest.Email, opt => opt.MapFrom(src => src.Email))
+                .ForMember(dest => dest.CustomerId, opt => opt.MapFrom(src => src.CustomerId))
+                .ForMember(dest => dest.FirstName, opt => opt.MapFrom(src => src.FirstName))
+                .ForMember(dest => dest.LastName, opt => opt.MapFrom(src => src.LastName))
+                .ForAllOtherMembers(opt => opt.Ignore());
+
+            this.CreateMap<LoginPipelineArgs, User>()
```
Comment (larionovadaria) [inline src/Foundation/Commerce/code/Models/Mappings/UserProfile.cs]: Is that a `.ReverseMap()` ?

File: src/Foundation/Commerce/code/ModelMappers/Users/UserMapper.cs (hunk 7 above)
Comment (yuya-yuliya) [reply src/Foundation/Commerce/code/ModelMappers/Users/UserMapper.cs]: But we can have many profiles for one mapper

File: src/Foundation/Commerce/code/Models/Mappings/UserProfile.cs (hunk 10 above)
Comment (yuya-yuliya) [reply src/Foundation/Commerce/code/Models/Mappings/UserProfile.cs]: We need to ignore some fields in both directions, so `ReverseMap` not work correctly

File: src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/GetCommerceUserProcessor.cs (hunk 3 above)
Comment (yuya-yuliya) [reply src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/GetCommerceUserProcessor.cs]: For extensibility it's better to separate abort reasons

File: src/Foundation/Commerce/code/ModelMappers/Users/UserMapper.cs (hunk 7 above)
Comment (yuya-yuliya) [reply src/Foundation/Commerce/code/ModelMappers/Users/UserMapper.cs]: We'll create task for this

File: src/Foundation/Commerce/code/Models/Mappings/UserProfile.cs (hunk 10 above)
Comment (larionovadaria) [reply src/Foundation/Commerce/code/Models/Mappings/UserProfile.cs]: ok

File: src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/GetCommerceUserProcessor.cs (hunk 3 above)
Comment (larionovadaria) [reply src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/GetCommerceUserProcessor.cs]: Use `AddMessage` / `GetMessages`

File: src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/IdentifyCommerceUserProcessor.cs [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+            this.commerceTrackingService = commerceTrackingService;
+            this.logService = logService;
+        }
+
+        public override void Process(LoginPipelineArgs args)
+        {
+            Assert.ArgumentNotNull(args, nameof(args));
+
+            try
```
Comment (larionovadaria) [inline src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/IdentifyCommerceUserProcessor.cs]: Good, but need to move this try/catch + addMessage + errorLog to abstract `DoProcessSafe` and refactor usages.

File: src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/MergeCartsProcessor.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+            if (args.AnonymousContactId != null)
+            {
+                var result = this.cartService.MergeCarts(args.AnonymousContactId);
+
+                if (!result.Success)
+                {
+                    foreach (var resultError in result.Errors)
+                    {
+                        args.AddMessage(resultError, PipelineMessageType.Warning);
```
Comment (larionovadaria) [inline src/Foundation/Commerce/code/Infrastructure/Pipelines/Login/MergeCartsProcessor.cs]: If merge cart fails it's definitely an error, we can't proceed with that, we have to `AbortPipeline`.

**Review by ghost — CHANGES_REQUESTED** [review]

**Pull request merged**